File: Prog1/prog1/altklausuren/14ws/tmp/assignment13/NodesEqualToParent.java
/*
Compile: 
javac -cp .:prog1javalib.jar NodesEqualToParent.java
Run:
java -cp .:prog1javalib.jar NodesEqualToParent
Compile and run:
javac -cp .:prog1javalib.jar NodesEqualToParent.java && java -cp .:prog1javalib.jar NodesEqualToParent
*/

import prog1.base.Base;

class Node {
    public int value;
    public Node left, right;

    public Node(Node left, int value, Node right) {
        this.left = left;
        this.value = value;
        this.right = right;
    }

    public Node(int value) {
        this.value = value;
        left = null;
        right = null;
    }

}

class Tree {

    private Node root = null;
    
    public Tree() { }

    public Tree(Node root) {
        this.root = root;
    }

    public String toString() {
        return toStringRec(root);
    }

    private String toStringRec(Node n) {
        if (n == null) {
            return "E";
        } else {
            return "Node(" + toStringRec(n.left) 
                    + ", " + n.value + ", "
                    + toStringRec(n.right) + ")";
        }
    }
    
    /** Returns the number of nodes that have the same value as their parent node. 
      * May need a helper method. */
    public int numberOfNodesThatAreEqualToTheirParent() {
        return 0; // todo: implement
    }
    
}
    
public class NodesEqualToParent {       

    public static Node n(int value) {
        return new Node(value);
    }
    
    public static Node n(Node left, int value, Node right) {
        return new Node(left, value, right);
    }

    public static void test() {
        Tree t = new Tree();
        Base.checkExpect(t.numberOfNodesThatAreEqualToTheirParent(), 0);
        
        t = new Tree(n(100));
        Base.checkExpect(t.numberOfNodesThatAreEqualToTheirParent(), 0);
        
        t = new Tree(n(n(99), 100, n(100)));
        Base.checkExpect(t.numberOfNodesThatAreEqualToTheirParent(), 1);
        
        t = new Tree(n(n(100), 100, n(100)));
        Base.checkExpect(t.numberOfNodesThatAreEqualToTheirParent(), 2);
        
        t = new Tree(n(n(101), 100, n(100)));
        Base.checkExpect(t.numberOfNodesThatAreEqualToTheirParent(), 1);
        
        t = new Tree(n(n(n(25), 50, n(75)), 100, n(n(125), 150, n(175))));
        Base.checkExpect(t.numberOfNodesThatAreEqualToTheirParent(), 0);
        
        t = new Tree(n(n(n(0), 0, n(0)), 
                    0, 
                    n(n(0), 0, n(0))));
        Base.checkExpect(t.numberOfNodesThatAreEqualToTheirParent(), 6);
        
        t = new Tree(n(null, 100, n(100)));
        Base.checkExpect(t.numberOfNodesThatAreEqualToTheirParent(), 1);
        
        t = new Tree(n(null, 100, n(99)));
        Base.checkExpect(t.numberOfNodesThatAreEqualToTheirParent(), 0);
        
        t = new Tree(n(n(100), 100, null));
        Base.checkExpect(t.numberOfNodesThatAreEqualToTheirParent(), 1);
        
        Base.summary();
    }

    public static void main(String[] args) {
        test();
    }

}

File: Prog1/prog1/altklausuren/altKlausurNEW/assignment14/musterLösung/nodes_equal_to_parent.c
/*
Compile: make nodes_equal_to_parent
Run: ./nodes_equal_to_parent
Compile & run: make nodes_equal_to_parent && ./nodes_equal_to_parent
*/

#include "base.h"

typedef struct Node{
	int value;
	struct Node* left;
	struct Node* right;
}Node;

Node* new_node(Node* left, int value, Node* right){
	Node* node = xmalloc(sizeof(Node));
	node->left = left;
	node->value = value;
	node->right = right;
	return node;
}

void free_node(Node * node){
	if(node){
		free_node(node->left);
		free_node(node->right);
		node->left = NULL;
		node->right = NULL;
		free(node);
	}
}

void print_node(Node* node) {
	if (node == NULL) {
		printf("E");
	} else {
		printf("Node(");
		print_node(node->left);
		printf(", %d, ", node->value);
		print_node(node->right);
		printf(")");
	}
}

//todo
int number_of_nodes_that_are_equal_to_their_parent_nodes(Node* node) {
	int i = 0;
	if (node->left != NULL) {
		if (node->left->value == node->value) i++;
		i += number_of_nodes_that_are_equal_to_their_parent_nodes(node->left);
	}
	if (node->right != NULL) {
		if (node->right->value == node->value) i++;
		i += number_of_nodes_that_are_equal_to_their_parent_nodes(node->right);
	}
	return i;
}

typedef struct{
	Node* root;
}Tree;

Tree* new_tree(Node* root){
	Tree* tree = xmalloc(sizeof(Tree));
	tree->root = root;
	return tree;
}

void free_tree(Tree* tree){
	if(tree){
		free_node(tree->root);
		tree->root = NULL;
		free(tree);
	}
}

/** Returns the number of nodes that have the same value as their parent node. 
      * May need a helper method. */
int number_of_nodes_that_are_equal_to_their_parent(Tree* tree) {
	if (tree->root == NULL) {
		return 0;
	} else {
		return number_of_nodes_that_are_equal_to_their_parent_nodes(tree->root);
	}
}
void print_tree(Tree* tree){
	printf("Tree: ");
	if(tree)
		print_node(tree->root);
	printf("\n");
}
	
	

void test() {
	Tree* t = new_tree(NULL);
	test_equal_i(number_of_nodes_that_are_equal_to_their_parent(t), 0);
	free_tree(t);
		
	t = new_tree(new_node(NULL, 100, NULL));
	test_equal_i(number_of_nodes_that_are_equal_to_their_parent(t), 0);
	free_tree(t);
	
	t = new_tree(new_node(new_node(NULL, 99, NULL), 100, new_node(NULL, 100, NULL)));
	test_equal_i(number_of_nodes_that_are_equal_to_their_parent(t), 1);
	free_tree(t);
	
	t = new_tree(new_node(new_node(NULL, 100, NULL), 100, new_node(NULL, 100, NULL)));
	test_equal_i(number_of_nodes_that_are_equal_to_their_parent(t), 2);
	free_tree(t);
		
	t = new_tree(new_node(new_node(NULL, 101, NULL), 100, new_node(NULL, 100, NULL)));
	test_equal_i(number_of_nodes_that_are_equal_to_their_parent(t), 1);
	free_tree(t);
	
	t = new_tree(new_node(new_node(new_node(NULL, 25, NULL), 50, new_node(NULL,75, NULL)), 100, new_node(new_node(NULL,125, NULL), 150, new_node(NULL, 175, NULL))));
	test_equal_i(number_of_nodes_that_are_equal_to_their_parent(t), 0);
	free_tree(t);

	t = new_tree(new_node(new_node(new_node(NULL, 0, NULL), 0, new_node(NULL, 0, NULL)), 0, new_node(new_node(NULL, 0, NULL), 0, new_node(NULL, 0, NULL))));
	test_equal_i(number_of_nodes_that_are_equal_to_their_parent(t), 6);
	free_tree(t);
		
	t = new_tree(new_node(NULL, 100, new_node(NULL, 100, NULL)));
	test_equal_i(number_of_nodes_that_are_equal_to_their_parent(t), 1);
	free_tree(t);
	
	t = new_tree(new_node(NULL, 100, new_node(NULL, 99, NULL)));
	test_equal_i(number_of_nodes_that_are_equal_to_their_parent(t), 0);
	free_tree(t);
		
	t = new_tree(new_node(new_node(NULL, 100, NULL), 100, NULL));
	test_equal_i(number_of_nodes_that_are_equal_to_their_parent(t), 1);
	free_tree(t);

}	
	
int main(void) {
	base_init();
    base_set_memory_check(true);
	test();
	return 0;
}